### PR TITLE
[SPARK-49819] Disable CollapseProject for correlated subqueries in projection over aggregate correctly

### PR DESCRIPTION
**Pull request by n-young-db**
### What changes were proposed in this pull request?

CollapseProject should block collapsing with an aggregate if any correlated subquery is present. There are other correlated subqueries that are not ScalarSubquery that are not accounted for here.

### Why are the changes needed?

Availability issue.

### Does this PR introduce _any_ user-facing change?

Previously failing queries will not fail anymore.

### How was this patch tested?

UT.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala [hunk 1]
```diff
@@ -2154,6 +2154,24 @@ class SubquerySuite extends QueryTest
     }
   }
 
+  test("SPARK-49819: Do not collapse projects with exist subqueries") {
+    withTempView("t1", "t2") {
```
Comment (bersprockets) [inline sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala:2158]: With `"t1", "t2"` or with `"v"`?

Comment (n-young-db) [reply sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: Fixed!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -1232,11 +1233,8 @@ object CollapseProject extends Rule[LogicalPlan] with AliasHelper {
    * in aggregate if they are also part of the grouping expressions. Otherwise the plan
    * after subquery rewrite will not be valid.
    */
-  private def canCollapseAggregate(p: Project, a: Aggregate): Boolean = {
-    p.projectList.forall(_.collect {
-      case s: ScalarSubquery if s.outerAttrs.nonEmpty => s
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1237]: good catch! not only scalar subquery expression can appear in the SELECT list.

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala [hunk 3]
```diff
@@ -2154,6 +2154,24 @@ class SubquerySuite extends QueryTest
     }
   }
 
+  test("SPARK-49819: Do not collapse projects with exist subqueries") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala:2157]:
```suggestion
  test("SPARK-49819: Do not collapse projects with exist subqueries ") {
```

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala [hunk 4]
```diff
@@ -2154,6 +2154,24 @@ class SubquerySuite extends QueryTest
     }
   }
 
+  test("SPARK-49819: Do not collapse projects with exist subqueries ") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala:2157]:
```suggestion
  test("SPARK-49819: Do not collapse projects with exist subqueries") {
```

**Comment by cloud-fan** [comment]
LGTM. I've re-triggered CI and will merge it when tests pass

**Comment by cloud-fan** [comment]
thanks, merging to master/3.5!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 5]
```diff
@@ -1232,11 +1233,8 @@ object CollapseProject extends Rule[LogicalPlan] with AliasHelper {
...
-    }.isEmpty)
-  }
+  private def canCollapseAggregate(p: Project, a: Aggregate): Boolean =
+    !p.projectList.exists(hasCorrelatedSubquery)
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1237]: Shall we remove the `a: Aggregate`?

**Review by beliefer — APPROVED** [review]

**Pull request closed without merging**